### PR TITLE
EOS-28602: Validate the UT for M0_OIF_SKIP_LAYOUT when enabled forcefully

### PR DESCRIPTION
**Pull request by imvenkip**
# Problem Statement
Validate the UT for M0_OIF_SKIP_LAYOUT when enabled forcefully

# Design
Added two UT's to validate for M0_OIF_SKIP_LAYOUT.
 1. ut_dix_namei_ops_dist_oi_flags - Index operations
 2. ut_dix_record_ops_dist_oi_flags - Record operations

# Coding
   Checklist for Author
-  [X] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [X] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [X] JIRA number/GitHub Issue added to PR
- [X] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [X] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: motr/ut/idx_dix.c [hunk 1]
```diff
@@ -399,12 +416,18 @@ static void ut_dix_namei_ops(bool dist)
 
 static void ut_dix_namei_ops_dist(void)
 {
-	ut_dix_namei_ops(true);
+	ut_dix_namei_ops(true, 0);
+}
+
+static void ut_dix_namei_ops_dist_oi_flags(void)
```
Comment (vidyadhar-pinglikar) [inline motr/ut/idx_dix.c:422]: may be we can add one more ut, where both oif_skip_layout and oif_crow flag  are set.

Comment (imvenkip) [reply motr/ut/idx_dix.c]: Added tow more UT's with skip_layout and crow flag.

**Review by vidyadhar-pinglikar — APPROVED** [review]

File: motr/ut/idx_dix.c [hunk 2]
```diff
@@ -399,12 +416,24 @@ static void ut_dix_namei_ops(bool dist)
 
 static void ut_dix_namei_ops_dist(void)
 {
-	ut_dix_namei_ops(true);
+	ut_dix_namei_ops(true, 0);
+}
+
+static void ut_dix_namei_ops_dist_oi_flags_skip_layout(void)
```
Comment (madhavemuri) [inline motr/ut/idx_dix.c:422]: Rename ut_dix_namei_ops_dist_oi_flags_skip_layout() to  ut_dix_namei_ops_skip_layout()
And do the same for other corresponding API's .

File: motr/ut/idx_dix.c [hunk 3]
```diff
@@ -417,30 +446,37 @@ static uint64_t dix_val(uint64_t i)
...
+	int                  rc;
+	int                 *rcs;
+	struct m0_op_common *oc;
+	struct m0_op_idx    *oi;
 
 	idx_dix_ut_init();
 	general_ifid_fill(&ifid, dist);
 	m0_container_init(&realm, NULL, &M0_UBER_REALM, ut_m0c);
 	m0_idx_init(&idx, &realm.co_realm, (struct m0_uint128 *)&ifid);
+	if(flags & M0_OIF_SKIP_LAYOUT)
```
Comment (madhavemuri) [inline motr/ut/idx_dix.c:471]: Add a space after if .

File: motr/ut/idx_dix.c [hunk 4]
```diff
@@ -353,17 +365,22 @@ static void ut_dix_namei_ops(bool dist)
...
 	M0_UT_ASSERT(rc == 0);
-	rc = m0_idx_op(&idx, M0_IC_LIST, &keys, NULL, rcs, 0,
+	rc = m0_idx_op(&idx, M0_IC_LIST, &keys, NULL, rcs, flags,
 			      &op);
 	M0_UT_ASSERT(rc == 0);
 	m0_op_launch(&op, 1);
 	rc = m0_op_wait(op, M0_BITS(M0_OS_STABLE), WAIT_TIMEOUT);
 	M0_UT_ASSERT(rc == 0);
-	M0_UT_ASSERT(rcs[0] == 0);
+	if(flags & M0_OIF_SKIP_LAYOUT) {
```
Comment (madhavemuri) [inline motr/ut/idx_dix.c:374]: Add a space after if.

File: motr/ut/idx_dix.c (hunk 2 above)
Comment (imvenkip) [reply motr/ut/idx_dix.c:422]: Done.

File: motr/ut/idx_dix.c (hunk 3 above)
Comment (imvenkip) [reply motr/ut/idx_dix.c:471]: Done

File: motr/ut/idx_dix.c (hunk 4 above)
Comment (imvenkip) [reply motr/ut/idx_dix.c:374]: Done.

File: motr/idx_dix.c [hunk 5]
```diff
@@ -1262,10 +1262,11 @@ M0_INTERNAL int m0__idx_cancel(struct m0_op_idx *oi)
 }
 
 static void dix_set_idx_flags(struct m0_op_idx *oi)
-{
+{ 
 	if (ENABLE_DTM0)
 		oi->oi_flags |= M0_OIF_SKIP_LAYOUT;
-	else
+
+	if(!(oi->oi_flags & M0_OIF_SKIP_LAYOUT))
```
Comment (madhavemuri) [inline motr/idx_dix.c:1269]: Add space after if ().

File: motr/ut/idx_dix.c [hunk 6]
```diff
@@ -302,25 +302,32 @@ static void ut_dix_namei_ops_cancel_non_dist(void)
...
+	struct m0_op_idx    *oi;
+	
 	idx_dix_ut_init();
 	m0_container_init(&realm, NULL, &M0_UBER_REALM, ut_m0c);
-
 	general_ifid_fill(&ifid, dist);
 	/* Create the index. */
 	m0_idx_init(&idx, &realm.co_realm, (struct m0_uint128 *)&ifid);
+
+	if(flags & M0_OIF_SKIP_LAYOUT)
```
Comment (madhavemuri) [inline motr/ut/idx_dix.c:324]: Add space after if ().

File: motr/ut/idx_dix.c [hunk 7]
```diff
@@ -329,8 +336,13 @@ static void ut_dix_namei_ops(bool dist)
 
 	/* Create an index with the same fid once more => -EEXIST. */
 	m0_idx_init(&idup, &realm.co_realm, (struct m0_uint128 *)&ifid);
+	if(flags & M0_OIF_SKIP_LAYOUT)
```
Comment (madhavemuri) [inline motr/ut/idx_dix.c:339]: Add space after if ().

**Review by madhavemuri — APPROVED** [review]

**Comment by ivan-alekhin** [comment]
Was this patch (as well as the previous one -- https://github.com/Seagate/cortx-motr/pull/1301) tested with `--enable-dtm0` option? There are several UTs (idx-dix-mt:dtm0*) that could be used to verify the code hidden behind ENABLE_DTM0.

Also, could you please describe the relation between M0_OIF_SKIP_LAYOUT, M0_ENF_META, COF_SKIP_LAYOUT, M0_OIF_CROW, COF_CROW,  ENABLE_DTM0 flags w.r.t to the use-cases and the corresponding state machines (for example, Motr client sm or dixreq sm) somewhere here https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/822313181/CAS+Service+DTM+Changes? It would be really helpful for us to see what kinds of behavior we can expect here. These flags belong to different domains (cas, dix, entity, operation, dtm), so it would be good to have the whole picture. cc @sg-shankar

**Pull request merged**